src/autoscaler/metricsforwarder/server/custom_metrics_handlers.go

```diff
@@ -0,0 +1,193 @@
+				Code:    "Authorization-Failure-Error",
+				Message: "Basic authorization credential does not match"})
+			return
+		}
+		// Valid credentials found in DB so update the cache
+		mh.credentialCache.Set(appID, credentials, mh.cacheTTL)
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		mh.logger.Error("error-reading-request-body", err, lager.Data{"body": r.Body})
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error reading custom metrics request body"})
+		return
+	}
+	var metricsConsumer *models.MetricsConsumer
+	err = json.Unmarshal(body, &metricsConsumer)
+	if err != nil {
+		mh.logger.Error("error-unmarshaling-metrics", err, lager.Data{"body": r.Body})
+		handlers.WriteJSONResponse(w, http.StatusInternalServerError, models.ErrorResponse{
+			Code:    "Interal-Server-Error",
+			Message: "Error unmarshaling custom metrics request body"})
+		return
+	}
+	isValidRequest, err := mh.validateCustomMetricTypes(appID, metricsConsumer)
+	if !isValidRequest && err != nil {
+		mh.logger.Error("failed-validating-metrictypes", err, lager.Data{"metrics": metricsConsumer})
+		handlers.WriteJSONResponse(w, http.StatusBadRequest, models.ErrorResponse{
+			Code:    "Bad-Request",
+			Message: "Error validating custom metrics type"})
+		return
+	}
+
+	metrics := mh.getMetrics(appID, metricsConsumer)
+
+	if len(metrics) <= 0 {
+		mh.logger.Debug("failed-parsing-custom-metrics-request-body", lager.Data{"metrics": metrics})
+		handlers.WriteJSONResponse(w, http.StatusBadRequest, models.ErrorResponse{
+			Code:    "Bad-Request",
+			Message: "Error parsing request body"})
+		return
+	}
+
+	mh.logger.Debug("custom-metrics-parsed-successfully", lager.Data{"metrics": metrics})
+	for _, metric := range metrics {
+		mh.metricForwarder.EmitMetric(metric)
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
+func (mh *CustomMetricsHandler) validateCredentials(username string, usernameHash string, password string, passwordHash string) bool {
+	isUsernameAuthenticated := bcrypt.CompareHashAndPassword([]byte(usernameHash), []byte(username))
+	isPasswordAuthenticated := bcrypt.CompareHashAndPassword([]byte(passwordHash), []byte(password))
+	if isPasswordAuthenticated == nil && isUsernameAuthenticated == nil { // password matching successfull
+		return true
+	}
+	mh.logger.Debug("failed-to-authorize-credentials")
+	return false
+}
+
+func (mh *CustomMetricsHandler) validateCustomMetricTypes(appGUID string, metricsConsumer *models.MetricsConsumer) (bool, error) {
+	standardMetricsTypes := make(map[string]struct{})
+	standardMetricsTypes[models.MetricNameMemoryUsed] = struct{}{}
+	standardMetricsTypes[models.MetricNameMemoryUtil] = struct{}{}
+	standardMetricsTypes[models.MetricNameThroughput] = struct{}{}
+	standardMetricsTypes[models.MetricNameResponseTime] = struct{}{}
+
+	scalingPolicy, err := mh.policyDB.GetAppPolicy(appGUID)
```

Comment:
we need to fetch AppPolicy for each request and parse the allowedMetricTypeSet here. 

How about caching the allowedMetricTypeSet as well, and set a reasonable expire TTL., i.e. 5 minutes. 
Just as the credential cache,   
if cache expired, fetch it from policy db
if not found,  fetch it from policy db
if the incoming metric type is invalid, fetch it from policy db to get the latest policy and revalidate
if the incoming metric type is valid previously, but not valid anymore due to the policy changes  and cache entry doesn't expire,  then just forward the metric to loggregator.   Then, in metric collector,  we can add an implementation to discard the staled metrics .

Reply:
Updated after adding cache for metric types